src/plugins/chart_expressions/expression_partition_vis/public/utils/layers/sort_predicate.ts

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { ArrayEntry } from '@elastic/charts';
+import { Datatable } from '../../../../../../../src/plugins/expressions';
+import { BucketColumns, ChartTypes, PartitionVisParams } from '../../../common/types';
+
+type SortFn = (([name1, node1]: ArrayEntry, [name2, node2]: ArrayEntry) => number) | undefined;
+
+type SortPredicateDefaultFn = (
+  visData: Datatable,
+  columns: Array<Partial<BucketColumns>>
+) => SortFn;
+
+type SortPredicatePieDonutFn = (visParams: PartitionVisParams) => SortFn;
+
+type SortPredicatePureFn = () => SortFn;
+
+export const extractUniqTermsMap = (dataTable: Datatable, columnId: string) =>
+  [...new Set(dataTable.rows.map((item) => item[columnId]))].reduce(
+    (acc, item, index) => ({
+      ...acc,
+      [item]: index,
+    }),
+    {}
+  );
+
+const sortPredicateSaveSourceOrder: SortPredicatePureFn =
+  () =>
+  ([, node1], [, node2]) => {
```

Comment:
I didn't test, but is `Predicate.DataIndex` from the elastic-charts plugin doing this as well? Used hereL https://github.com/elastic/kibana/blob/5423daffcf3b2fa48accd391fc7762a872f022d7/src/plugins/chart_expressions/expression_partition_vis/public/components/chart_split.tsx#L39